### PR TITLE
(API) Cluster Health report unassigned_primary_shards (#111727)

### LINKED ISSUE #111727: (API) Cluster Health report unassigned_primary_shards
### Description

👋 howdy, team!

Will you kindly consider having the Allocation [Cluster Health API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-health.html) also report `unassigned_primary_shards` on top of `unassigned_shards` to better summarize a `status:red`? 

E.g. under `-` I would suspect a node-level/higher issue but under `+` I would suspect an index-level. 

```diff
{
  "cluster_name": "abea12cb73c74d1e994e3d23cff1340c",
  "status": "red",
  "timed_out": false,
  "number_of_nodes": 15,
  "number_of_data_nodes": 10,
  "active_primary_shards": 1895,
  "active_shards": 2294,
  "relocating_shards": 0,
  "initializing_shards": 2,
  "unassigned_shards": 62,
-  "unassigned_primary_shards": 62,
+  "unassigned_primary_shards": 1,
  "delayed_unassigned_shards": 0,
  "number_of_pending_tasks": 0,
  "number_of_in_flight_fetch": 0,
  "task_max_waiting_in_queue_millis": 0,
  "active_shards_percent_as_number": 97.28583545377438
}
```

Obviously next step is still run Allocation Explain, but this  would let you quick check amount of shards you potentially have to review to resolve `red` > `yellow` status before you start manually reading through CAT Shards to find a shard to check.

### PR DESCRIPTION
**Pull request by lukewhiting**
This PR adds a count of currently unassigned primary shards to both the `/_cat/health` and `/_cluster/health` endpoints. This is to aid cluster administrators in estimating the time remaining for a cluster to go from RED to YELLOW status as per enchancement request #111727.

Tests and doc updates are in place with this PR and manual testing with `./gradlew run` has been conducted on the endpoints to ensure correct output.

## Known Limitations
* Testing
  * Due to limitations in the YAML REST test framework skip functionality, YAML REST tests for this endpoint are disabled when running a mixed version cluster by using a cluster version number synthetic feature to skip when any member of the cluster is not at a version greater than when this change is due to be introduced

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:82]: I don't immediately see why we need this check? If the number of unassigned primary shards is 0, we can still just add it, right?

File: server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:224]:
```suggestion
        builder.field(UNASSIGNED_PRIMARY_SHARDS, getUnassignedPrimaryShards());
```

File: server/src/main/java/org/elasticsearch/rest/action/cat/RestHealthAction.java
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/rest/action/cat/RestHealthAction.java:74]: I don't think this aligns with our Checkstyle configuration. I don't know if you already got the `Adapter for Eclipse Code Formatter` plugin working already, but in the meanwhile, you can run `gradle spotlessApply` to format all code in compliance with our Checkstyle configuration.

File: server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java
Comment (lukewhiting) [reply server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java]: Good spot :-) Fixed

File: server/src/main/java/org/elasticsearch/rest/action/cat/RestHealthAction.java
Comment (lukewhiting) [reply server/src/main/java/org/elasticsearch/rest/action/cat/RestHealthAction.java]: Have got this all setup now and thanks for the tip on spotlessApply!

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml
Comment (gmarouli) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:106]: Are these tests ever executed? According to the docs:
```
Note that `skip` with `capabilities` or `cluster_features` will skip the test if *any* node in the cluster
has the feature or capability. `requires` will only run the test if *all* of the nodes in the cluster
have the feature or capability.
```
I understand this in the following way:
- Mixed clusters with the current version will be skipped.
- This would only run if the all nodes of the cluster are on version less than 8.15.0

Considering the above I am not sure that these test are worth having. Clusters with mixed versions less than 8.15.0 should have been tested by their own tests. I would like to have mixed cluster tests but I would like one of the nodes to be on the current version.

What do you think?

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:47]: I would prefer to keep these two at the top of the class. I am not sure it's convention necessarily, but if you see the majority of the `BaseRestHandlers` have them on top, and it's very helpful to find this metadata at the top of the class.

I understand this is not how we structure classes in general, but in this case it is very helpful considering the nature of the class. Does this make sense to you?

I do not feel so strongly about `allowSystemIndexAccessByDefault`.

File: server/src/test/java/org/elasticsearch/cluster/routing/RoutingTableGenerator.java
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/cluster/routing/RoutingTableGenerator.java:137]: If I am not mistaken, here we changed the semantics of the `unassigned`. In the original version of the code it used to count all unassigned shards and now it is counting only the non-primaries. Was that intentional?

File: server/src/test/java/org/elasticsearch/cluster/health/ClusterIndexHealthTests.java
Comment (nielsbauman) [inline server/src/test/java/org/elasticsearch/cluster/health/ClusterIndexHealthTests.java:149]: This is obviously not your doing, but I'm not a big fan of implementations like these of `mutateInstance`. I prefer implementations like:
https://github.com/elastic/elasticsearch/blob/00f2b9a779adcd52c8172367e6c63132393c8cf9/server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java#L85-L85
because they're much more compact and readable (IMO).

I'll leave it up to you if you want to refactor that in this PR or you keep it like this -- if you even agree, which isn't required of course.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:42]: Can you revert all the changes here that aren't necessary? I don't see a strong reason to move these things around, so then I'd rather keep them as is. It looks like the only thing you really changed is adding the `supportedCapabilities` method, right?

**Review by gmarouli — CHANGES_REQUESTED** [review]
Thank you for picking this up @lukewhiting great work! I added a few small comments on the code, let me know what you think.

Furthermore I would like to bring up a bwc concern. 

Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

I see at least two paths forward:

1. We accept that, the trade-off of fixing this might not be worth it.
2. We use node features for bwc to ensure that when the unassigned primary shard is returned it's always accurate.

What do you think?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml
Comment (nielsbauman) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:19]: I won't ask you to revert these changes in this PR, but I personally try to avoid making formatting changes in files like these because there is no enforced format here - and reducing the number of changed lines makes it easier to review (not just the PR, but also the commit in case someone wants to look at this commit in the future).

**Review by nielsbauman — CHANGES_REQUESTED** [review]
This is great work @lukewhiting! Impressive first PR :)

Most of your changes LGTM, I just left a few small comments.

**Comment by nielsbauman** [comment]
@gmarouli
> Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

@lukewhiting and I also talked about that. I said I was fas in favor of this approach and your option 1 (we accept this the way it is) because this is only relevant in mixed clusters and thus temporarily. My initial thought was that I think it's reasonable that that count isn't accurate when the cluster is in the process of upgrading. On second thought, I guess one could also argue that if there are unassigned primary shards _because_ of the in-progress cluster upgrade, this new count might be valuable. However, I still think it's reasonable to return inaccurate counts during the cluster upgrade.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml
Comment (lukewhiting) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:106]: I think you are right... With the capabilities API I was intending to test the case of a mixed cluster when you hit an older node as a coordinator but as that API does not work correctly in this case and we had to switch to the features api, I think the test is now redundant.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml
Comment (lukewhiting) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:2]: Given your later comment about the other tests in this file, I'm going to revent the name back to the original name as your suggestion and totally remove the "without" tests.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java
Comment (lukewhiting) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:42]: Sure. My IDE got a bit enthusiastic with a reformat and applied the default "Static methods go at the top" rule to this class but I agree with you and @gmarouli that in the case of rest actions, it makes sense to have the route at the top.

File: server/src/test/java/org/elasticsearch/cluster/health/ClusterIndexHealthTests.java
Comment (lukewhiting) [reply server/src/test/java/org/elasticsearch/cluster/health/ClusterIndexHealthTests.java:149]: I agree it's a much better pattern with the vars and the pattern matching instead of a legacy switch.

I might even go one step further and suggest using a builder defined in an anonymous class inside the test. That would further help contain and simplify the mutation method.

That said, I'm going to open a seperate PR for that I think. This one is already quite large and I'm hesitant to introduce more changes.

**Comment by lukewhiting** [comment]
@gmarouli 
> Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

As @nielsbauman mentioned I did initially implemented this so that any not sending the new field through would cause it to be hidden on the rest output however we ended up deciding the additional complexity wasn't worth it for this being a temporary state.

Happy to be convinced otherwise though as I don't really know enough to have a strong opinion yet 😅

**Comment by gmarouli** [comment]
> As @nielsbauman mentioned I did initially implemented this so that any node not sending the new field through would cause it to be hidden on the rest output however we ended up deciding the additional complexity wasn't worth it for this being a temporary state.

I guess it depends on the usage of the feature. As @nielsbauman said, if people rely on it then it might be worth it. 

Maybe an idea is to document this. Shall we add a note in the docs that this value will be partially calculated if there nodes in the cluster with versions prior to 8.16, and if a user wants the accurate number they should use the health API?

**Comment by lukewhiting** [comment]
@nielsbauman @gmarouli 

I have added a warning to the docs about inaccurate counts in mixed clusters: https://github.com/elastic/elasticsearch/pull/112024/commits/73e2e06a7319182b2e62894164184f0caab87d34

As we're not 100% sure on the right action here, lets hold off merging and I'll bring it up at the weekly to see what the wider team think.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml
Comment (lukewhiting) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml]: Tests removed in latest commit

**Comment by lukewhiting** [comment]
@elasticsearchmachine buildkite test this

**Comment by lukewhiting** [comment]
Consensus from the team weekly was that the field being incorrect during cluster upgrade should not be an issue as best practice recommendation is that the master node is upgraded last and this endpoint runs on the master node.

With that in mind, the DM team are happy to proceed with the current logic.

File: docs/reference/cluster/health.asciidoc
Comment (gmarouli) [inline docs/reference/cluster/health.asciidoc:135]: A small suggestion to be inline with the language in the docs. After a quick look I saw that we do not use the prefix `v`, we just use the version number.
 
```suggestion
    (integer) The number of shards that are primary but not allocated. **Note**: This number may be lower than the true value if your cluster contains nodes running a version prior to 8.16. For a more accurate count in this scenario, please use the <<cluster-health,cluster health API>>.
```

**Review by gmarouli — APPROVED** [review]
LGTM, I added a small comment about how to describe the version in the docs. Apart from that you are good to go 🚀 Congrats, first PR 🥳 !

File: docs/reference/cluster/health.asciidoc
Comment (lukewhiting) [reply docs/reference/cluster/health.asciidoc]: Sure thing. Have pushed that change :-)

**Review by nielsbauman — APPROVED** [review]
LGTM! Great work on your first PR!! :rocket:

**Pull request merged**
